File: udp/send_local/python/udp_local_client.py
#!/usr/bin/env python3
import socket
from argparse import ArgumentParser
from datetime import datetime

MAX_BYTES = 65535

def client(port):
    sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
    text = 'The time is {}'.format(datetime.now())
    data = text.encode('ascii')
    sock.sendto(data, ('127.0.0.1', port))
    print('The OS assigned me the address {}'.format(sock.getsockname()))
    data, address = sock.recvfrom(MAX_BYTES)  # Danger! See Chapter 2
    text = data.decode('ascii')
    print('The server {} replied {!r}'.format(address, text))

if __name__ == '__main__':
    parser = ArgumentParser(description='Send and receive UDP locally')
    parser.add_argument('--port', '-p', type=int, default=1060, help='UDP port (default 1060)')
    args = parser.parse_args()

    print("args: {}".format(args))

    client(args.port)

File: udp/send_local/cxx/udp_local_client.cpp
#include <cstring>
#include <string>
#include <sstream>
#include <iostream>
#include <gflags/gflags.h>

#include "unp.hpp"

#define MAX_BYTES 65535

std::string datetime_now()
{
    char buf[255];
    memset(buf, 0, sizeof(buf));
    time_t t = time(0);
    strftime(buf, sizeof(buf), "%Y-%m-%d %H:%M:%S", localtime(&t));
    return buf;
}

void client(uint16_t port) {
    Socket sock(AF_INET, SOCK_DGRAM);
    auto text = "The time is " + datetime_now();
    sock.SendTo(text, SocketAddress("127.0.0.1", port));
    std::cout << "The OS assigned me the address " << sock.Getsockname() << std::endl;
    SocketAddress address;
    std::tie(text, address) = sock.RecvFrom(MAX_BYTES);
    std::cout << "The server " << address << " replied " << text << std::endl;
    sock.Close();
}

std::string usage(const char* prog) {
    std::ostringstream os;
    os << "Send and receive UDP locally\n\n"
        << "Usage: " << prog << " [--help] [--port PORT]";
    return os.str();
}

DEFINE_int32(port, 1060, "TCP port number");

int main(int argc, char *argv[]) {
    gflags::SetUsageMessage(usage(argv[0]));
    gflags::ParseCommandLineFlags(&argc, &argv, true);

    client(FLAGS_port);

    return 0;
}

